src/sagemaker_xgboost_container/distributed_gpu/distributed_gpu_training.py

```diff
@@ -0,0 +1,119 @@
+    scheduler_host_ip = get_host_ip(scheduler_host)
+
+    scheduler_address = f"{scheduler_host_ip}:{SCHEDULER_PORT}"
+    scheduler_conn_string = f"tcp://{scheduler_address}"
+    is_scheduler_host = current_host == scheduler_host
+
+    start_daemons_in_current_instance(scheduler_conn_string, is_scheduler_host)
+    time.sleep(WAIT_TIME_DAEMONS_STARTUP_SEC)
+
+    total_num_workers = len(sm_hosts) * num_gpus
+
+    # We only need to submit the job from one instance.
+    if is_scheduler_host:
+        with Client(scheduler_address) as client:
+            # We ensure that all workers are present before proceeding.
+            client.wait_for_workers(total_num_workers, WAIT_FOR_ALL_WORKERS_TIMEOUT_SEC)
+
+            logging.info("Starting to read training data...")
+            watchlist = []
+            try:
+                X_train, y_train = load_data_into_memory(client, train_path, content_type)
+
+                dtrain = xgb.dask.DaskDMatrix(client, X_train, y_train)
+
+                # Log train data dimension for sanity check.
+                train_num_rows, train_num_cols = get_dataframe_dimensions(X_train)
+                logging.info(f"Train features matrix has {train_num_rows} rows and {train_num_cols} columns")
+
+                watchlist.append((dtrain, "train"))
+
+                if validation_path is not None:
+                    X_valid, y_valid = load_data_into_memory(client, validation_path, content_type)
+                    dvalid = xgb.dask.DaskDMatrix(client, X_valid, y_valid)
+                    watchlist.append((dvalid, "validation"))
+
+            except Exception as e:
+                raise exc.UserError(f"Failed to load data with exception:\n{e}")
+
+            logging.info("Data load complete. Starting training...")
+
+            try:
+                output = xgb.dask.train(
+                    client, hyperparameters, dtrain, num_boost_round=hyperparameters["num_round"], evals=watchlist
+                )
+                booster = output["booster"]
+
+                logging.info("Training complete. Saving model...")
+                booster.save_model(os.path.join(model_dir, MODEL_NAME))
+            except Exception as e:
+                exception_prefix = "XGB train call failed with exception"
+                raise exc.AlgorithmError(f"{exception_prefix}:\n {str(e)}")
+
+            logging.info("Terminating cluster...")
+
+    else:
+        # Do not exit till the job is done.
+        while True:
+            scheduler = (scheduler_host_ip, int(SCHEDULER_PORT))
+            alive_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            alive_check = alive_socket.connect_ex(scheduler)
+            if alive_check == 0:
+                pass
+            else:
+                logging.info("Received a shutdown signal from scheduler. Exiting...")
+                break
+            alive_socket.close()
+            time.sleep(WORKER_STAY_ALIVE_CHECK_FREQ_SEC)
```

Comment:
If we want to keep a socket alive and call connect_ex we'll have to handle "already connected" 106 errors.

Reply:
Hmm, so if I'm following correctly on line 117 there's an execution path where we want to break out of the while loop, but keep `alive_socket` open? When would we close the socket then?

Reply:
In this case it wouldn't matter whether we close the connection or not, since that'd be the end of the program. This `else` portion is purely just to keep the programs in non-scheduler hosts from exiting early.
Your comment raises a good question though, whether opening a new connection every 2s will cause any issues, especially when we get upto a large number of hosts. For now I'd prefer to move forward with a longer check period, and we can revisit this as follow-up.

Reply:
OK, in that case I think this should work then and aligns with the original functionality:

```python3
         while True:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as alive_socket:
                scheduler = (scheduler_host_ip, int(SCHEDULER_PORT))
                alive_check = alive_socket.connect_ex(scheduler)
                if alive_check != 0:
                    logging.info("Received a shutdown signal from scheduler. Exiting...")
                    break
            time.sleep(WORKER_STAY_ALIVE_CHECK_FREQ_SEC)
```